### PR TITLE
Add dtslint to CI

### LINKED ISSUE #11: Incorporate linting in CI
[`namoscato:ci...ci-lint`](https://github.com/namoscato/google-api-typings-generator/compare/ci...namoscato:ci-lint) prototypes build / linting CI, and [here is an example run](https://github.com/namoscato/google-api-typings-generator/commit/8a5bb2f0dd033e54134964b1cb47238d2a29eb3b/checks?check_suite_id=379045990) which is currently failing.

Thoughts on this approach and how it could relate to generating these type definitions automatically longterm?

### PR DESCRIPTION
**Pull request by namoscato**
Resolves #11 

## Details

Runs [`dtslint`](https://github.com/microsoft/dtslint) on generated projects in parallel via [`npm-run-all`](https://github.com/mysticatea/npm-run-all).

## Open Questions

- [x] Should we exclude failing projects for now?
- [x] Should we just run this on pull requests?

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,3 +23,25 @@ jobs:
...
+      with:
+        node-version: ${{ matrix.node-version }}
+    - name: npm install, build, and lint
+      run: |
+        npm ci
+        npm start
+        npm run lint
+      env:
+        CI: true
+        GAPI_MAX_PARALLEL: 3
```
Comment (namoscato) [inline .github/workflows/ci.yml]: I did some rough tests to come up with this number, i.e. see:

* https://github.com/namoscato/google-api-typings-generator/runs/374746191
* https://github.com/namoscato/google-api-typings-generator/runs/374752209

File: package.json [hunk 2]
```diff
@@ -2,8 +2,10 @@
   "name": "google-api-typings-generator",
   "description": "Google API Typescript typings definitions generator",
   "scripts": {
+    "dtslint": "dtslint",
+    "lint": "ts-node bin/lint.ts ./types",
     "start": "ts-node -T src/cli.ts --out ./types",
-    "test": "mocha tests/tests.ts --require ts-node/register"
+    "test": "tsc --noEmit && mocha tests/tests.ts --require ts-node/register"
```
Comment (namoscato) [inline package.json]: This adds type checking as part of the test script.

File: package.json [hunk 3]
```diff
@@ -15,13 +17,14 @@
...
+    "dtslint": "^2.0.3",
     "jasmine": "^3.5.0",
-    "mocha": "^6.2.2"
+    "mocha": "^6.2.2",
+    "npm-run-all": "^4.1.5"
   },
   "dependencies": {
     "commander": "^4.0.1",
     "dot": "^1.1.3",
-    "dtslint": "^2.0.3",
```
Comment (namoscato) [inline package.json]: Thinking about this, should we just make _all_ dependencies dev given this project is not really intended to be pulled in via a package manager?

Comment (Maxim-Mazurok) [reply package.json]: Well, maybe we will publish it to npm someday, so that users could generate fresh typings using `npx google-api-typings-generator ...`. 
I'm not sure, does `npx` pull `devDependencies`? I don't think so.

File: bin/lint.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { readdirSync } from 'fs';
+import { join } from 'path';
+
+const MAX_PARALLEL = Number(process.env.GAPI_MAX_PARALLEL) || 1;
+
+process.stdout.setMaxListeners(MAX_PARALLEL);
+process.stderr.setMaxListeners(1 + MAX_PARALLEL);
+
+const path = process.argv[2];
+const runAll = require('npm-run-all');
```
Comment (Maxim-Mazurok) [inline bin/lint.ts]: Let's prefer `import * as runAll from 'npm-run-all';` ?
And it probably makes sense to put it in the beginning of the file, together with other imports.

**Comment by Maxim-Mazurok** [comment]
Thanks for great PR!

> Should we exclude failing projects for now?

Only dialogflow typings are failing. They seem to have JSdoc used in [library](https://github.com/googleapis/nodejs-dialogflow). I've tried to use our typings with their lib and they are not compatible.
A couple of examples taken from lib samples:
```ts
const sessionClient: gapi.client.dialogflow.SessionsResource = new dialogflow.SessionsClient();
const sessionPath = sessionClient.sessionPath(projectId, sessionId);
```
This gives error, because `.sessionPath()` is not in our typings, but it's lib function.

```ts
const request = {
  session: sessionPath,
  queryInput: {
    text: {
      // The query to send to the dialogflow agent
      text: 'hello',
      // The language used by the client (en-US)
      languageCode: 'en-US',
    },
  },
};

// Send request and log result
const responses = await sessionClient.detectIntent(request);
```
Here, you can see that they pass `queryInput` to `detectIntent()`. But according to our typings, `detectIntent()` doesn't accept `queryInput`, instead you have to pass it as a part of `resource` parameter (`GoogleCloudDialogflowV2DetectIntentRequest`).

Long story short, I think there's not much we can do right now about `dialogflow`, so we should ignore it, by adding to `excludedApi` array.

> Should we just run this on pull requests?

Sounds good enough for now. What else are you thinking? Running it as a part of git-hook using [husky](https://www.npmjs.com/package/husky) before committing/pushing, maybe?

File: bin/lint.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { readdirSync } from 'fs';
+import { join } from 'path';
+
+const MAX_PARALLEL = Number(process.env.GAPI_MAX_PARALLEL) || 1;
```
Comment (Maxim-Mazurok) [inline bin/lint.ts]: Would be awesome to add [dotenv](https://www.npmjs.com/package/dotenv) support for running linting locally also in parallel.

**Comment by Maxim-Mazurok** [comment]
Also, found another DialogFlow lib, which is in TS already, so there's no need in typings: https://github.com/googleapis/google-api-nodejs-client/tree/master/src/apis/dialogflow

File: bin/lint.ts (hunk 5 above)
Comment (namoscato) [reply bin/lint.ts]: Yeah, this will also work:

```
$ GAPI_MAX_PARALLEL=2 npm run lint
```

Will that suffice for now?

File: bin/lint.ts (hunk 4 above)
Comment (namoscato) [reply bin/lint.ts]: Agreed. I was having issues with the `import` syntax:

```ts
import runAll from 'npm-run-all';
```

for some unexplainable reason:

```
runAll([scripts.shift()], options) // run first synchronously to install TypeScript
      ^
TypeError: npm_run_all_1.default is not a function
    at Object.<anonymous> (/Users/namoscato/dev/git/google-api-typings-generator/bin/lint.ts:28:7)
```

but I can look into it.

File: package.json (hunk 3 above)
Comment (namoscato) [reply package.json]: > I'm not sure, does `npx` pull `devDependencies`?

Yeah, if this project was installed as another project's dependency, `devDependencies` would be excluded. I'm still not sure I see the use case, but I'm fine leaving the rest of these as-is.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (namoscato) [inline .github/workflows/ci.yml]: > Should we just run this on pull requests?

Actually, per [Pull request events for forked repositories](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#pull-request-events-for-forked-repositories), `push` events are not triggered in fork contexts, so I think we'll at least want these running on `pull_request`.

> Running it as a part of git-hook using [husky](https://www.npmjs.com/package/husky) before committing/pushing, maybe?

The linting is relatively taxing; I don't think we're going to want to be running it locally but happy to chat about it more.

File: bin/lint.ts (hunk 5 above)
Comment (Maxim-Mazurok) [reply bin/lint.ts]: Sure, add this snippet to the `README.md` then, please

File: bin/lint.ts (hunk 4 above)
Comment (Maxim-Mazurok) [reply bin/lint.ts]: Yeah, that's probably because they don't have default export, so 
```js
import * as runAll from 'npm-run-all';
```
will work (`* as` part)

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Maxim-Mazurok) [reply .github/workflows/ci.yml]: > `push` events are not triggered in fork contexts, so I think we'll

I can only see this about **pull** events, not **push**: "no pull request events occur on the forked repository". Am I missing something?...

> The linting is relatively taxing

Totally agree here, no need to enforce local linting for now

File: bin/lint.ts (hunk 4 above)
Comment (namoscato) [reply bin/lint.ts]: _Right_ – thanks!

File: .github/workflows/ci.yml (hunk 1 above)
Comment (namoscato) [reply .github/workflows/ci.yml]: > I can only see this about **pull** events, not **push**: "no pull request events occur on the forked repository". Am I missing something?...

Well, the fact that the `push` workflows have not been triggered within the context of _this_ pull request is telling. I moved the linting job to `pull_request`, so 🤞 that starts to work once this is merged to `master`?

Comment (namoscato) [reply .github/workflows/ci.yml]: That did the trick! ✅
